library/ssl_tls.c

```diff
@@ -1908,25 +1943,38 @@ STATIC int ssl_decrypt_buf( mbedtls_ssl_context *ssl,
                     transform->ivlen );
         }
 #endif
-        if( rec->data_len == 0 )
-        {
-            MBEDTLS_SSL_DEBUG_MSG( 1, ( "message doesn't contain padding byte" ) );
-            return( MBEDTLS_ERR_SSL_INVALID_MAC );
-        }
 
+        /* Safe since data_len >= minlen + maclen + 1, so after having
+         * subtracted at most minlen and maclen up to this point,
+         * data_len > 0. */
         padlen = data[rec->data_len - 1];
-#if defined(MBEDTLS_SSL_DEBUG_ALL)
-        if( rec->data_len < padlen + 1 )
+
+        if( auth_done == 1 )
         {
-            MBEDTLS_SSL_DEBUG_MSG( 1, ( "msglen (%d) < 1 + padlen (%d)",
-                                        rec->data_len, padlen ) );
+            correct *= ( rec->data_len >= padlen + 1 );
+            padlen  *= ( rec->data_len >= padlen + 1 );
         }
+        else
+        {
+#if defined(MBEDTLS_SSL_DEBUG_ALL)
+            if( rec->data_len < transform->maclen + padlen + 1 )
+            {
+                MBEDTLS_SSL_DEBUG_MSG( 1, ( "msglen (%d) < maclen (%d) + padlen (%d)",
```

Comment:
The value of `padlen` is the value of the padding length byte at this point, so the entire padding consists of `padlen + 1` bytes. I'm inclined to leave the debug message as it, because to me the splitting of the padding into 'padding bytes' and 'padding length byte' makes it less readable.